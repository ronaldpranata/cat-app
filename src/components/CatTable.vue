<template>
  <div class='table'>
    <div class='row header'>
      <div class='column' v-for='(item, index) in headers' :key='index'>
        {{ item }}
      </div>
    </div>
    <div class='row' v-for='(item, index) in body' :key='index'>
      <div class='column' v-for='(value, key) in item' :key='key'>
        <span v-if="key=='image'">
          <img :src="value" width="60" />
        </span>
         <span v-else>
          {{ value }}
        </span>
      </div>
    </div>
  </div>
</template>

<script>
import { defineComponent } from 'vue';
import { useI18n } from 'vue-i18n';

export default defineComponent({
  name: 'CatTable',
  props: {
    headers: Array,
    body: Array,
  },
  setup() {
    const { t } = useI18n();
    return { t };
  },
});
</script>
<style scoped lang='scss'>
.table {
  max-width: 500px;
  margin: auto;
  .row {
    display: flex;
    flex-direction: row;
  }
  .column {
    display: flex;
    flex:1;
    flex-direction: column;
  }
}
</style>
