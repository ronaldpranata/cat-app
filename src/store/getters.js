export default {
  getCatBreedColumn() {
    return ['Image', 'Name'];
  },
  getCatBreed(state) {
    return state.catBreeds;
  },
  getCatDetail(state) {
    return state.catDetail;
  },
  getOriginMap(state) {
    return state.catOriginMap;
  },
};
