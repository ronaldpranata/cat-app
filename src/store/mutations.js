export default {
  SET_CAT_BREEDS: (state, data) => {
    state.catBreeds = data;
  },
  SET_MAP_LOCATION: (state, data) => {
    state.catOriginMap = data;
  },
};
