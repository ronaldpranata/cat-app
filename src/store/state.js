export default () => ({
  catBreeds: [],
  catOriginMap: null,
});
