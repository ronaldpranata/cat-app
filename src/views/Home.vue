<template>
  <div class="home">
    <h1>{{ t("message") }}</h1>
    <div class="error" v-if="error" >{{error}}</div>
    <div class="loading" v-if="loading">Loading.....<br/><br/><br/></div>
    <CatTable
      v-else-if="!error"
      :body="tableBody"
      :headers="tableHeader"
      @gotoDetail="gotoDetail"
    />
  </div>
</template>

<script>
// @ is an alias to /src
import * as _ from 'lodash';
import { useI18n } from 'vue-i18n';
import { computed } from 'vue';
import { useRouter } from 'vue-router';
import CatTable from '@/components/CatTable.vue';
import useCatBreed from '@/composables/useCatBreed';

export default {
  name: 'Home',
  components: {
    CatTable,
  },
  setup() {
    const { t } = useI18n();
    const router = useRouter();
    const { cats, loading, error } = useCatBreed();
    const tableHeader = [t('image'), t('name')];
    const tableBody = computed(() => _.map(cats.value, (o) => ({
      image: o?.image?.url,
      name: o?.name,
    })));

    async function gotoDetail(name) {
      await router.push({ path: `/catDetail/${name}` });
    }
    return {
      t, tableHeader, tableBody, gotoDetail, loading, error,
    };
  },
};
</script>
