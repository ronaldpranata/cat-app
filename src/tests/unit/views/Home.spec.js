import { mount } from '@vue/test-utils';
import { createStore } from 'vuex';
import Home from '@/views/Home.vue';
import CatTable from '@/components/CatTable.vue';
import mockCatBreeds from '@/__mock__/mock';
import { createI18n } from 'vue-i18n';
import state from '@/store/state';
import mutations from '@/store/mutations';
import getters from '@/store/getters';
import {
  GET_CAT_BREEDS,
} from '@/store/action_types';

const i18n = createI18n({
  // vue-i18n options here ...
});

const actions = {
  async [GET_CAT_BREEDS]({ commit }) {
    try {
      const response = await new Promise((resolve) => {
        resolve(mockCatBreeds);
      });
      commit('SET_CAT_BREEDS', response?.data);
      return response?.data;
    } catch (error) {
      throw new Error(error);
    }
  },
};
const store = createStore({
  state,
  actions,
  mutations,
  getters,
});

const factory = () => mount(Home, {
  global: {
    plugins: [i18n, store],
  },
});

describe('Home', () => {
  test('renders a CatTable', () => {
    const wrapper = factory();
    const component = wrapper.findComponent(CatTable);
    expect(component.exists()).toBeTruthy();
  });

  test('mounts properly', () => {
    const wrapper = factory();
    expect(wrapper.vm).toBeTruthy();
  });

  test('renders properly', () => {
    const wrapper = factory();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
