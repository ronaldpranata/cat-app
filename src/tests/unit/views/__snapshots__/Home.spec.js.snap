// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders properly 1`] = `
<div class="home">
  <h1>message</h1>
  <!--v-if-->
  <div class="table">
    <div class="row header">
      <div class="column">Image</div>
      <div class="column">Name</div>
    </div>
  </div>
</div>
`;
